src/Microsoft.AspNetCore.Razor.Language/Legacy/MarkupParser.cs

```diff
@@ -0,0 +1,916 @@
+
+                ParseMarkupNodes(builder, ParseMode.Markup);
+                AcceptMarkerTokenIfNecessary();
+                builder.Add(OutputAsMarkupLiteral());
+
+                var markup = SyntaxFactory.MarkupBlock(builder.ToList());
+
+                return SyntaxFactory.RazorDocument(markup);
+            }
+        }
+
+        private void ParseMarkupNodes(
+            in SyntaxListBuilder<RazorSyntaxNode> builder,
+            ParseMode mode,
+            Func<SyntaxToken, bool> stopCondition = null)
+        {
+            stopCondition = stopCondition ?? (token => false);
+            while (!EndOfFile && !stopCondition(CurrentToken))
+            {
+                ParseMarkupNode(builder, mode);
+            }
+        }
+
+        private void ParseMarkupNode(in SyntaxListBuilder<RazorSyntaxNode> builder, ParseMode mode)
+        {
+            switch (GetParserState(mode))
+            {
+                case ParserState.MarkupText:
+                    ParseMarkupText(builder);
+                    break;
+                case ParserState.Tag:
+                    ParseMarkupElement(builder);
+                    break;
+                case ParserState.SpecialTag:
+                    ParseSpecialTag(builder);
+                    break;
+                case ParserState.XmlPI:
+                    ParseXmlPI(builder);
+                    break;
+                case ParserState.CData:
+                    ParseCData(builder);
+                    break;
+                case ParserState.MarkupComment:
+                    ParseMarkupComment(builder);
+                    break;
+                case ParserState.RazorComment:
+                    ParseRazorCommentWithLeadingAndTrailingWhitespace(builder);
+                    break;
+                case ParserState.DoubleTransition:
+                    ParseDoubleTransition(builder);
+                    break;
+                case ParserState.CodeTransition:
+                    ParseCodeTransition(builder);
+                    break;
+                case ParserState.Misc:
+                    ParseMisc(builder);
+                    break;
+                case ParserState.Unknown:
+                    AcceptAndMoveNext();
+                    break;
+                case ParserState.EOF:
+                    builder.Add(OutputAsMarkupLiteral());
+                    break;
+            }
+        }
+
+        private void ParseMarkupText(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            AcceptAndMoveNext();
+        }
+
+        private void ParseMarkupElement(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            Assert(SyntaxKind.OpenAngle);
+
+            // Output already accepted tokens if any.
+            builder.Add(OutputAsMarkupLiteral());
+
+            if (!NextIs(SyntaxKind.ForwardSlash))
+            {
+                // Parsing a start tag
+                var startTag = ParseStartTag(out var expectsBody);
```

Comment:
I'd remove the `expectsBody` and have that be part of the signature of the tag block. Aka `startTag.Mode` or something like that. One thing to note, it makes sense to assign a tag mode in the creation of the tag so we don't bake too much knowledge into the tag block.

WIth the below comments, this entire block could look like:
```C#
var startTag = ParseStartTag();
builder.Add(startTag);

if (startTag.Mode == MarkupTagMode.Void || startTag.Mode == MarkupTagMode.SelfClosing)
{
    // Does not expect a body
    return;
}

var body = ParseTagBody(startTag); // Would return a <missing> instead of null
builder.Add(endTag)

var endTag = ParseEndTag(startTag); // Can return missing if it doesn't find the end tag or log an error or w/e
builder.Add(endTag)
```

Another option here would be to always parse the tag body and always parse the end tag and have those methods always return <missing> in the case of an unexpecting body.